### PR TITLE
Provide bindings/swig iff we are using swig_matlab

### LINKED ISSUE #2957: Drake should allow building without swig_matlab
From what I understand, only a couple externals are allowed to be mandatory (`Eigen`, `googletest`).  However, the code in `bindings/swig/CMakeLists.txt` appears to have a hard dependency on the `swig_matlab` external being present anytime python is present.  This seems wrong, especially given the gargantuan size of that external.

### PR DESCRIPTION
**Pull request by jwnimmer-tri**
Provide `bindings/swig` if and only if we are using `swig_matlab`, though in an admittedly somewhat brittle way.

Fixes #2957.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/2961)

### CODE REVIEW

**Comment by jwnimmer-tri** [comment]
Probably shouldn't merge until we've passed some special opt-in mac + matlab build.  Will do that later.

**Comment by jwnimmer-tri** [comment]
This failing in Jenkins even in normal builds.  DNR while I try to get it working.

**Comment by jwnimmer-tri** [comment]
@drake-jenkins-bot mac-clang-experimental-matlab please
_Fingers crossed._

**Comment by jwnimmer-tri** [comment]
@jamiesnape or @mwoehlke-kitware or @billhoffman could I get a volunteer to self-assign this and do a feature review, please?  No particular hurry, I'd just like to get it on someone's plate.

**Comment by mwoehlke-kitware** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/bindings/CMakeLists.txt, line 2 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-KNm24BaTsW0R37stRAV:-KNm24BaTsW0R37stRAW:-413499390) ([raw file](https://github.com/robotlocomotion/drake/blob/d367048c5737976d8dc428c4c9291380c3402ae6/drake/bindings/CMakeLists.txt#L2)):*
> ```CMake
> if(NOT WIN32)
>   set(CMAKE_MODULE_PATH
> ```

Okay, but I've always preferred `list(APPEND CMAKE_MODULE_PATH ${CMAKE_INSTALL_PREFIX}/share/cmake)`

---

*[drake/bindings/CMakeLists.txt, line 5 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-KNm2E_yt-7E8KrUIb3h:-KNm2E_zOaOQ3XULxJvp:1526607562) ([raw file](https://github.com/robotlocomotion/drake/blob/d367048c5737976d8dc428c4c9291380c3402ae6/drake/bindings/CMakeLists.txt#L5)):*
> ```CMake
>     ${CMAKE_INSTALL_PREFIX}/share/cmake
>     ${CMAKE_MODULE_PATH})
>   if(EXISTS "${CMAKE_INSTALL_PREFIX}/bin/swig")
> ```

`find_program(SWIG_EXECUTABLE swig PATHS ${CMAKE_INSTALL_PREFIX}/bin NO_DEFAULT_PATH)`

---

*[drake/bindings/CMakeLists.txt, line 8 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-KNm37S-a-2ciIXIt0xx:-KNm37S-a-2ciIXIt0xy:-1203765250) ([raw file](https://github.com/robotlocomotion/drake/blob/d367048c5737976d8dc428c4c9291380c3402ae6/drake/bindings/CMakeLists.txt#L8)):*
> ```CMake
>     set(SWIG_EXECUTABLE ${CMAKE_INSTALL_PREFIX}/bin/swig)
>   endif()
>   # N.B. Ubuntu Xenial provides 3.0.8, so saying 3.0.9 here (which is
> ```

This may stop working at some point, or on other Linux distros. Maybe this whole block should be guarded by `if(SWIG_EXECUTABLE)`, since that will only pick up one built by the superbuild?

Conversely, with that change, I wonder if the other changes (i.e. moving the call to `find_package(SWIG)` and making it not `REQUIRED`) are even necessary?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2961)*

**Comment by jwnimmer-tri** [comment]
Thanks, and good questions.  I will do another round of edits this afternoon -- anytime I change anything, it takes a while to re-test locally in several different configs, before pushing.  _Sigh_.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-:-KNm45qCP8L1mHT4pom_:580240315)*

**Comment by jwnimmer-tri** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/bindings/CMakeLists.txt, line 2 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-KNm24BaTsW0R37stRAV:-KNmdqujdI6Kbp7pfdfL:-937898019) ([raw file](https://github.com/robotlocomotion/drake/blob/d367048c5737976d8dc428c4c9291380c3402ae6/drake/bindings/CMakeLists.txt#L2)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary>
> Okay, but I've always preferred `list(APPEND CMAKE_MODULE_PATH ${CMAKE_INSTALL_PREFIX}/share/cmake)`
</details>
Done (in `bindings/swig/CMakeLists.txt`; I returned this to its original home).

---

*[drake/bindings/CMakeLists.txt, line 5 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-KNm2E_yt-7E8KrUIb3h:-KNme5jMO9-TOzASGvB3:-1810025239) ([raw file](https://github.com/robotlocomotion/drake/blob/d367048c5737976d8dc428c4c9291380c3402ae6/drake/bindings/CMakeLists.txt#L5)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary>
> `find_program(SWIG_EXECUTABLE swig PATHS ${CMAKE_INSTALL_PREFIX}/bin NO_DEFAULT_PATH)`
</details>
Done. (It turns out I needed more `NO_`... goop to get it to stop finding things in `/usr/...`.)

---

*[drake/bindings/CMakeLists.txt, line 8 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-KNm37S-a-2ciIXIt0xx:-KNmeQ_H8FOP1XHWkWrX:649405780) ([raw file](https://github.com/robotlocomotion/drake/blob/d367048c5737976d8dc428c4c9291380c3402ae6/drake/bindings/CMakeLists.txt#L8)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary>
> This may stop working at some point, or on other Linux distros. Maybe this whole block should be guarded by `if(SWIG_EXECUTABLE)`, since that will only pick up one built by the superbuild?
> 
> Conversely, with that change, I wonder if the other changes (i.e. moving the call to `find_package(SWIG)` and making it not `REQUIRED`) are even necessary?
</details>
That does seem cleaner and more robust.  Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2961)*

**Comment by mwoehlke-kitware** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/bindings/CMakeLists.txt, line 2 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-KNm24BaTsW0R37stRAV:-KNmz9PneiD8EG7pNFEI:591690111) ([raw file](https://github.com/robotlocomotion/drake/blob/d367048c5737976d8dc428c4c9291380c3402ae6/drake/bindings/CMakeLists.txt#L2)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Done (in `bindings/swig/CMakeLists.txt`; I returned this to its original home).
</details>
Thanks! :smile:

---

*[drake/bindings/CMakeLists.txt, line 5 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-KNm2E_yt-7E8KrUIb3h:-KNmyj1uhgi1bCqCjUph:1100330946) ([raw file](https://github.com/robotlocomotion/drake/blob/d367048c5737976d8dc428c4c9291380c3402ae6/drake/bindings/CMakeLists.txt#L5)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Done. (It turns out I needed more `NO_`... goop to get it to stop finding things in `/usr/...`.)
</details>
Hmm... [the documentation](https://cmake.org/cmake/help/v3.5/command/find_program.html) seems to strongly suggest that only `NO_DEFAULT_PATH` should be necessary. Also, it seems to work for me. When you tested, did you delete the cache variable first?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2961)*

**Comment by jwnimmer-tri** [comment]
*[drake/bindings/CMakeLists.txt, line 5 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-KNm2E_yt-7E8KrUIb3h:-KNn1ry-Ip59ORnEVe50:649003743) ([raw file](https://github.com/robotlocomotion/drake/blob/d367048c5737976d8dc428c4c9291380c3402ae6/drake/bindings/CMakeLists.txt#L5)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary>
> Hmm... [the documentation](https://cmake.org/cmake/help/v3.5/command/find_program.html) seems to strongly suggest that only `NO_DEFAULT_PATH` should be necessary. Also, it seems to work for me. When you tested, did you delete the cache variable first?
</details>
Ah.  Right, while initially testing, I was not clearing the cache; even though I later did, I must have backed myself into some corner.  I will try to put back just the `NO_DEFAULT_PATH`.

BTW How is anyone supposed to test anything with CMake successfully?  If I edit the `CMakeLists.txt`, things get re-run but apparently the cache is sticky and preserves old, wrong values?  Is there a way to tell CMake to be less unsound?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2961)*

**Comment by jamiesnape** [comment]
@drake-jenkins-bot linux-clang-experimental-minimal please

**Comment by mwoehlke-kitware** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/bindings/CMakeLists.txt, line 5 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-KNm2E_yt-7E8KrUIb3h:-KNn3DzmjoXgSmm0GxU3:49713427) ([raw file](https://github.com/robotlocomotion/drake/blob/d367048c5737976d8dc428c4c9291380c3402ae6/drake/bindings/CMakeLists.txt#L5)):*
> Apparently the cache is sticky and preserves old, wrong values? Is there a way to tell CMake to be less unsound?

In general, the `find_` set of functions *need* to work this way; otherwise the user wouldn't be able to tell CMake where to find something it can't find in the default paths.

In this case, if you're sure you want to force the One True Value and not ever let the user override it, you could `unset(SWIG_EXECUTABLE CACHE)` before calling `find_program`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2961)*

**Comment by jwnimmer-tri** [comment]
*[drake/bindings/CMakeLists.txt, line 5 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-KNm2E_yt-7E8KrUIb3h:-KNn3ra74UAyFld-mR0C:1382002917) ([raw file](https://github.com/robotlocomotion/drake/blob/d367048c5737976d8dc428c4c9291380c3402ae6/drake/bindings/CMakeLists.txt#L5)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary>
> > Apparently the cache is sticky and preserves old, wrong values? Is there a way to tell CMake to be less unsound?
> 
> In general, the `find_` set of functions *need* to work this way; otherwise the user wouldn't be able to tell CMake where to find something it can't find in the default paths.
> 
> In this case, if you're sure you want to force the One True Value and not ever let the user override it, you could `unset(SWIG_EXECUTABLE CACHE)` before calling `find_program`.
</details>
Ah.  Because the cache is both a memoizing database used within CMake, and a user input system, CMake can't nuke it when `CMakeLists.txt` files change.  I guess that makes sense.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2961)*

**Comment by jwnimmer-tri** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/bindings/CMakeLists.txt, line 5 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-KNm2E_yt-7E8KrUIb3h:-KNn7kcr1VfOObBJZ1Rb:2142785329) ([raw file](https://github.com/robotlocomotion/drake/blob/d367048c5737976d8dc428c4c9291380c3402ae6/drake/bindings/CMakeLists.txt#L5)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Ah.  Because the cache is both a memoizing database used within CMake, and a user input system, CMake can't nuke it when `CMakeLists.txt` files change.  I guess that makes sense.
</details>
Hopefully better now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2961)*

**Comment by mwoehlke-kitware** [comment]
:lgtm:

@jwnimmer-tri, not sure from your last comment if you intended to add the `unset` or not... either way is fine by me.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-:-KNnFNE0yBFGDHfqpzu9:1733749986)*

**Comment by jwnimmer-tri** [comment]
I am okay leaving the `unset` absent.  It seems kinder to users who may want to override the SWIG_EXECUTABLE, e.g., if they have their own superbuild.

**Comment by jwnimmer-tri** [comment]
@drake-jenkins-bot mac-clang-experimental-matlab please
@drake-jenkins-bot linux-clang-experimental-minimal please
@david-german-tri for platform review (including any routing through Russ's test and approval).

**Comment by david-german-tri** [comment]
@jwnimmer-tri Looks good to me.  I will test it locally on Mac + Linux with MATLAB tomorrow (or possibly later tonight).

**Comment by jwnimmer-tri** [comment]
N.B. Even though `continuous-integration/jenkins/linux-clang-experimental-minimal` failed due to #2972, it still demonstrates that the absences `swig_matlab` no longer causes any problems.

**Comment by jwnimmer-tri** [comment]
@david-german-tri any glitches with this PR?

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-:-KNsVX506EN2dB1-4dAL:165091700)*

**Comment by david-german-tri** [comment]
Whoops, I forgot about this, sorry.  Testing now.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-:-KNseUgMUhqY0Ce5GASb:1632778327)*

**Comment by david-german-tri** [comment]
I am stymied by lack of MATLAB licenses.  :-(

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-:-KNsoDpDQ8Eu_p_XsHbm:-1771362846)*

**Comment by RussTedrake** [comment]
:lgtm:

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-:-KO4_NYKMoZ8CGAbUvpd:1417088721)*

**Comment by RussTedrake** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/bindings/CMakeLists.txt, line 8 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-KO4aL0lXzce8-5aR1zn:-KO4aL0m9j52EaPg1DNY:-344274039) ([raw file](https://github.com/robotlocomotion/drake/blob/4d8e8ec96a486ccfa05b1fb5941de2912030284b/drake/bindings/CMakeLists.txt#L8)):*
> ```CMake
>     PATHS ${CMAKE_INSTALL_PREFIX}/bin NO_DEFAULT_PATH)
>   if(SWIG_EXECUTABLE)
>     add_subdirectory(swig)
> ```

OK, but perhaps add a warning message explaining that the swig binary was not found so swig bindings will be disabled. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2961)*

**Comment by jwnimmer-tri** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/bindings/CMakeLists.txt, line 8 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-KO4aL0lXzce8-5aR1zn:-KO5ZjVrlkrox26tcBFF:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/5c09a63e9a9b615cf3020f178f724a764e3a20f3/drake/bindings/CMakeLists.txt#L8)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> OK, but perhaps add a warning message explaining that the swig binary was not found so swig bindings will be disabled. 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2961)*

**Comment by jwnimmer-tri** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/bindings/CMakeLists.txt, line 5 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-KNm2E_yt-7E8KrUIb3h:-KO5_7hhTPSSn7UIhc7j:611509846) ([raw file](https://github.com/robotlocomotion/drake/blob/d367048c5737976d8dc428c4c9291380c3402ae6/drake/bindings/CMakeLists.txt#L5)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Hopefully better now.
</details>
 @mwoehlke-kitware need your Ack on this discussion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2961)*

File: drake/bindings/CMakeLists.txt [hunk 1]
```diff
@@ -1,5 +1,15 @@
...
-  add_subdirectory(swig)
+  # Look for the swig provided by our superbuild.  If it doesn't exist,
+  # don't fall back to system swig because it may be the wrong version
+  # and likely does not support matlab.
+  find_program(SWIG_EXECUTABLE swig
+    PATHS ${CMAKE_INSTALL_PREFIX}/bin NO_DEFAULT_PATH)
+  if(SWIG_EXECUTABLE)
+    add_subdirectory(swig)
+  else()
+    message(WARNING "\n----- WARNING: "
```
Comment (jamiesnape) [inline drake/bindings/CMakeLists.txt]: CMake already prints "CMake Warning" before the message.

**Comment by jwnimmer-tri** [comment]
> CMake already prints "CMake Warning" before the message.

Please always use reviewable comments, never github diff comments.

--

Here is the output when `swig_matlab` is absent:
```
CMake Warning at bindings/CMakeLists.txt:10 (message):
  

  ----- WARNING: swig_matlab not found; disabling swig_matlab bindings.


-- Writing path utilities to /home/jwnimmer/jwnimmer-tri/drake-distro/build/minimal-debug/install/matlab
```

Given Russ's request, I felt that a minimal error message was not satisfactory.  A minimal build already has about 50 lines of warnings, so I wanted to make this one more prevalent.

**Comment by jamiesnape** [comment]
btw you may find CDash more useful than scanning logfiles, and it will regex on CMake Warning.

**Comment by mwoehlke-kitware** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/bindings/CMakeLists.txt, line 5 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2961#-KNm2E_yt-7E8KrUIb3h:-KO5s3OMxMtjupawb31Z:1690247421) ([raw file](https://github.com/robotlocomotion/drake/blob/d367048c5737976d8dc428c4c9291380c3402ae6/drake/bindings/CMakeLists.txt#L5)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
>  @mwoehlke-kitware need your Ack on this discussion.
</details>
Arf, ReviewBoard ate a comment here at some point I think (but not relevant any more).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2961)*

**Comment by jwnimmer-tri** [comment]
https://drake-jenkins.csail.mit.edu/job/linux-gcc-ninja-continuous-matlab-open-source/319/ is upset.  Investigating.

**Comment by jwnimmer-tri** [comment]
> Clone of 'https://github.com/RobotLocomotion/director.git' into submodule path 'externals/director' failed

So maybe just a one-time glitch.

**Comment by mwoehlke-kitware** [comment]
Yike, it's not one-time (many builds are blowing up), though I don't think this PR is the problem... lots of errors fetching from github...

**Comment by mwoehlke-kitware** [comment]
(FTR: It appears there was a problem with github.com. CI is back to normal now.)

**Pull request merged**